x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java

```diff
@@ -0,0 +1,376 @@
+        assert context.getAuthenticationToken() != null : "existing authentication token must not be null";
+        context.setHandleNullToken(false);  // already has a token, should not try null token
+        doAuthenticate(context, existingCredentialsAuthenticators, false, listener);
+    }
+
+    private void doAuthenticate(
+        Authenticator.Context context,
+        List<Authenticator> authenticators,
+        boolean shouldExtractCredentials,
+        ActionListener<Authentication> listener
+    ) {
+        final IteratingActionListener<Authenticator.Result, Authenticator> iteratingActionListener = new IteratingActionListener<>(
+            ActionListener.wrap(result -> {
+                if (result.getStatus() == Authenticator.Status.SUCCESS) {
+                    maybeLookupRunAsUser(context, result.getAuthentication(), listener);
+                } else {
+                    if (context.shouldHandleNullToken()) {
+                        handleNullToken(context, listener);
+                    } else {
+                        listener.onFailure(Exceptions.authenticationError("failed to authenticate", result.getException()));
+                    }
+                }
+            }, listener::onFailure),
+            getAuthenticatorConsumer(context, shouldExtractCredentials),
+            authenticators,
+            context.getThreadContext(),
+            Function.identity(),
+            result -> result.getStatus() == Authenticator.Status.UNSUCCESSFUL || result.getStatus() == Authenticator.Status.NOT_HANDLED);
+        iteratingActionListener.run();
+    }
+
+    private BiConsumer<Authenticator, ActionListener<Authenticator.Result>> getAuthenticatorConsumer(
+        Authenticator.Context context,
+        boolean shouldExtractCredentials
+    ) {
+        return (authenticator, listener) -> {
+            if (shouldExtractCredentials) {
+                final AuthenticationToken authenticationToken;
+                try {
+                    authenticationToken = authenticator.extractCredentials(context);
+                } catch (Exception e) {
+                    if (e instanceof ElasticsearchSecurityException) {
+                        listener.onFailure(e);
+                    } else { // other exceptions like illegal argument
+                        context.addUnsuccessfulMessage(authenticator.name() + ": " + e.getMessage());
+                        listener.onResponse(Authenticator.Result.unsuccessful(e.getMessage(), e));
+                    }
+                    return;
+                }
+                if (authenticationToken == null) {
+                    listener.onResponse(Authenticator.Result.notHandled());
+                    return;
+                }
+                context.addAuthenticationToken(authenticationToken);
+            }
+            context.setHandleNullToken(context.shouldHandleNullToken() && authenticator.canFollowedByNullTokenHandler());
```

Comment:
This "null token handling" logic is rather awkward. I did it to match the existing behaviour. But it can definitely be simplified once we agree to tighten the logic.